### PR TITLE
feat: scytl connector

### PR DESCRIPTION
**Pull request by agreenspan24**
[Scytl](https://www.scytl.com/resources-and-references/customers/u-s-elections/), or Clarity Elections, is a company that creates a tool for publishing election results in real-time. It's used in the U.S. by several states and over 800 counties, including Georgia, Colorado, Arkansas, and Dallas County, Texas.

For each participating election administrator, they publish a site with results that can be published on election night. Unfortunately, while that site contains downloadable data, that data is formatted in a complex way, making it difficult for developers to fetch election results. In general, their results either come zipped in either an unformatted text file or a complex XML document. Summary results come as a zipped CSV, but just contain top-line results. The JSON that powers the site results is even more complicated.

This connector provides methods to download the latest election results from their site and formats them into readable lists of dictionaries, which can easily be converted into a Parsons Table or Pandas dataframe.

Because this connector is can be useful for live reporting, it also contains a polling feature. As long as the class is instantiated, it will only fetch results that are new since the previous fetch of that method. To skip this feature, set force_update to true on any of the fetch methods.

Note: please do not confuse with [Clarity Campaign Labs](https://www.claritycampaigns.com/)

Example microsite: https://results.enr.clarityelections.com/CO/105975/web.276603/

### CODE REVIEW

**Comment by shaunagm** [comment]
Hi @agreenspan24! Thanks so much for this contribution. I haven't had a chance to go through the whole PR yet but I'm hoping to get to it and leave a review later this week.

I appreciate you having thorough tests for you connector! I am a little hesitant about the 25-ish test files though. Are they all necessary? Might they cause issues for long-term maintainability if there are formatting changes?

**Comment by agreenspan24** [comment]
> Hi @agreenspan24! Thanks so much for this contribution. I haven't had a chance to go through the whole PR yet but I'm hoping to get to it and leave a review later this week.
> 
> I appreciate you having thorough tests for you connector! I am a little hesitant about the 25-ish test files though. Are they all necessary? Might they cause issues for long-term maintainability if there are formatting changes?

Hi Shauna! Makes total sense - was trying to keep the 18 county files to match the real results, but I think you're right to reduce them. I cut out all but two and adjusted the tests - run a good bit faster now too!

For maintainability, those files are just direct copies of real election results as Scytl holds them. If they change their format, I'm hoping we can just drop newer versions of those files into the tests and adjust the parser accordingly. But let me know your thoughts!

File: docs/scytl.rst
Comment (shaunagm) [inline docs/scytl.rst:15]: The vast majority of our connectors (maybe all of them, other than this one?) require authentication. It seems like this doesn't because the data source is public? It would be useful to call that out so that a reader doesn't think that authentication is required but not documented. You can do something like
```

.. note::
  Authentication
    All endpoints for Scytl are public, and do not need authentication.
```

File: parsons/scytl/scytl.py
Comment (shaunagm) [inline parsons/scytl/scytl.py:74]: It's unusual for anything other than auth info to get passed in to a Parsons connector, although it certainly does happen, and it makes sense to pass in `state` and `election_id` here so they don't have to be passed in with `get_summary_results`, etc. 

That said, `county` in particular confuses me. How does the county parameter here fit with the counties passed in with `get_detailed_results_for_participating_counties`? Does this county parameter get used at any other point? It seems like the other two are always state level but maybe I'm misunderstanding. Does passing in the county on init mean that `get_summary_results` and `get_county_results` apply only to the county?

File: parsons/scytl/scytl.py
Comment (shaunagm) [inline parsons/scytl/scytl.py:87]: Can you document the private methods? This is not for the users, since they won't display in our docs, but very helpful for future maintainability/readability.

**Review by shaunagm — CHANGES_REQUESTED** [review]
Overall looks good. Mostly docs issues here.

**Comment by shaunagm** [comment]
@agreenspan24 I think your test fixes look good!

File: parsons/scytl/scytl.py
Comment (agreenspan24) [reply parsons/scytl/scytl.py]: Updated!

File: parsons/scytl/scytl.py
Comment (agreenspan24) [reply parsons/scytl/scytl.py:74]: Yeah, the setup is kinda confusing, would love your help navigating it. 

Essentially, they have a concept of an election (at the state or county level), which can have summary results and detailed results. The detailed results show results by county for state level elections, while they show precinct results for a county-level election.

State-level elections can also have sub-elections in each county, and link to each of those county elections, which each have summary and detailed results. So, a county-level election won't have participating counties and the method won't make sense.

The main reason I wanted to add the state and election_id at the class level instead of the method is so the class could track the previous version- I have a feeling people would use this for polling and I want to make that as easy as possible for them to see if results have updated.

One idea might be to split this up into a `StateScytlElection` and `CountyScytlElection` class, which would have some of the same implementation, but not all of it. What do you think?

File: parsons/scytl/scytl.py
Comment (alxmrs) [inline parsons/scytl/scytl.py:13]: constants should be IN_APP_CAPS.

File: parsons/scytl/scytl.py
Comment (alxmrs) [inline parsons/scytl/scytl.py:18]: Is the double slash (`//`) intentional?

File: parsons/scytl/scytl.py
Comment (alxmrs) [inline parsons/scytl/scytl.py:42]: This could be a dataclass.

File: parsons/scytl/scytl.py
Comment (alxmrs) [inline parsons/scytl/scytl.py:58]: Why the `object` is correct (and the convention of the codebase), it also is unnecessary.

File: parsons/scytl/scytl.py
Comment (alxmrs) [inline parsons/scytl/scytl.py:87]:
```suggestion
    def _parse_date_to_utc(self, input_dt: str) -> datetime.datetime:
```

File: parsons/scytl/scytl.py
Comment (alxmrs) [inline parsons/scytl/scytl.py:156]: Better to use a logger for the class. -- Are you sure you don't want to raise an error? Should the pipeline continue if it can't read the data?

File: parsons/scytl/scytl.py
Comment (alxmrs) [inline parsons/scytl/scytl.py:144]: Consider wrapping this in a context manager to automatically handle closing and flushing.

File: parsons/scytl/scytl.py
Comment (alxmrs) [inline parsons/scytl/scytl.py:148]: Is a flush prudent here?

File: parsons/scytl/scytl.py
Comment (alxmrs) [inline parsons/scytl/scytl.py:160]:
```
import typing as t

t.Dict[str, CountryDetails]
```

File: parsons/scytl/scytl.py
Comment (alxmrs) [inline parsons/scytl/scytl.py:210]:
```
import typing as t
-> t.List[t.Dict]:
```

Comment (alxmrs) [reply parsons/scytl/scytl.py:210]: `object` != `dict`

File: parsons/scytl/scytl.py
Comment (alxmrs) [inline parsons/scytl/scytl.py:239]: parens are redundant.

File: parsons/scytl/scytl.py
Comment (alxmrs) [inline parsons/scytl/scytl.py:273]:
```suggestion
                            precinct_turnout = precinct_dict.get(precinct_name, {})
```

File: parsons/scytl/scytl.py
Comment (alxmrs) [inline parsons/scytl/scytl.py:242]: It looks like a lot of precinct info doesn't make it to the final function return. Am I missing something?

File: parsons/scytl/scytl.py
Comment (alxmrs) [inline parsons/scytl/scytl.py:353]:
```suggestion
                            county_turnout = county_dict.get(county_name, {})
```

File: parsons/scytl/scytl.py
Comment (alxmrs) [inline parsons/scytl/scytl.py:376]: See typing notes above.

File: parsons/scytl/scytl.py
Comment (alxmrs) [inline parsons/scytl/scytl.py:408]: I think a list comprehension would be more readable here.

File: parsons/scytl/scytl.py
Comment (alxmrs) [inline parsons/scytl/scytl.py:405]: recommended: use context managers to ensure that these get cleaned up on function return.

File: parsons/scytl/scytl.py
Comment (alxmrs) [inline parsons/scytl/scytl.py:427]:
```
import typing as t
-> t.List[Dict]:
```

File: parsons/scytl/scytl.py
Comment (alxmrs) [inline parsons/scytl/scytl.py:559]:
```
-> t.List[t.Union[t.Dict, str]]:
```

File: test/test_scytl/test_scytl.py
Comment (alxmrs) [inline test/test_scytl/test_scytl.py:11]: `_DIR`

File: test/test_scytl/test_scytl.py
Comment (alxmrs) [inline test/test_scytl/test_scytl.py:30]: I find it hard to understand if the data is correct since it's hard to compare the source and destination data files in this diff.

File: test/test_scytl/test_scytl.py
Comment (alxmrs) [inline test/test_scytl/test_scytl.py:44]: Is it possible to move this to the `setUp` method?

File: test/test_scytl/test_scytl.py
Comment (alxmrs) [inline test/test_scytl/test_scytl.py:50]: prefer unittest asserts over the `assert` operation since they give more specific error messages.

File: test/test_scytl/test_scytl.py
Comment (alxmrs) [inline test/test_scytl/test_scytl.py:71]: I bet a lot of this could go into either the `setUp` method or a class test utility method.

**Review by alxmrs — COMMENTED** [review]
A few notes.

File: parsons/scytl/scytl.py
Comment (agreenspan24) [reply parsons/scytl/scytl.py:156]: Another interesting idea. If @shaunagm wants to start using the typing package, that works for me, but I've just added normal type annotations for all the functions.

File: parsons/scytl/scytl.py
Comment (agreenspan24) [reply parsons/scytl/scytl.py:238]: Good catch, might as well add the rest to the dictionary.

File: test/test_scytl/test_scytl.py
Comment (agreenspan24) [reply test/test_scytl/test_scytl.py]: Yeah, looks like it, with a bit of hackyness.

File: test/test_scytl/test_scytl.py
Comment (agreenspan24) [reply test/test_scytl/test_scytl.py]: Yeah that's a reasonable point. I think it's important to use the zip files since unzipping is a big part of what we're testing.

I guess it's somewhat helpful to just read the expected CSV and make sure all the data looks good and reasonable, but not ideal.

File: parsons/scytl/scytl.py
Comment (agreenspan24) [reply parsons/scytl/scytl.py]: I'm not really sure how context managers work, how would they help here?

File: parsons/scytl/scytl.py
Comment (agreenspan24) [reply parsons/scytl/scytl.py]: I'm not actually too familiar here with the underlying implementation. How would a flush work and be useful here? Do I just need to wrap the zipdata in a `with` statement?

File: parsons/scytl/scytl.py
Comment (agreenspan24) [reply parsons/scytl/scytl.py]: Letting it throw, catching in certain cases.

**Review by agreenspan24 — COMMENTED** [review]
@alxmrs Thanks for the review! Still pretty new to proper python conventions, so I appreciate the deeper look! A couple of ideas here sound interesting to me, like typing and dataclasses, but I don't want to introduce new patterns without approval from maintainers, so @shaunagm let us know your thoughts!

File: parsons/scytl/scytl.py
Comment (alxmrs) [reply parsons/scytl/scytl.py]: Sorry, I am using a really wonky name. By contextmanager, I mean a `with` expression. 

All IO in python inherits from the IOBase, which supports automatically closing files / freeing memory if you wrap things in `with` statements. 

https://docs.python.org/3/library/io.html#io.IOBase

Here, I think you have a memory leak, since `close()` is never called on `zipdata`.

File: parsons/scytl/scytl.py
Comment (alxmrs) [reply parsons/scytl/scytl.py]: While I agree that wrapping zipdata with `with` is needed (see above), you may also want to call flush. 

Flush tells the file object to actually write the data from the buffers, which usually happens when the file is closed. 
Usually, if you want to access the data sooner, it's prudent to explicitly call flush.

File: parsons/scytl/scytl.py
Comment (shaunagm) [reply parsons/scytl/scytl.py:42]: I'm of two minds about this. Turning it into a dataclass will improve readability for some, but may make the code harder to read and reason about for many who are newer to Python and unfamiliar with dataclasses. Perhaps we can switch to using a dataclass if you're up for it @agreenspan24 and put a code comment for readers linking to an explanation of dataclasses ie https://docs.python.org/3/library/dataclasses.html

**Comment by shaunagm** [comment]
> Still pretty new to proper python conventions, so I appreciate the deeper look! A couple of ideas here sound interesting to me, like typing and dataclasses, but I don't want to introduce new patterns without approval from maintainers, so @shaunagm let us know your thoughts!

Responded re: dataclasses inline. The question of typing is a bigger one. We currently don't require any sort of type annotations and certainly not any strict enforcement, but perhaps that's something we could at least ask of contributors going forward. We certainly don't have capacity to retroactively add typing to the existing codebase but incremental improvements are better than nothing.

Thanks for raising these questions @agreenspan24 & @alxmrs

File: parsons/scytl/scytl.py
Comment (agreenspan24) [reply parsons/scytl/scytl.py:44]: Added as a dataclass with comment

**Comment by agreenspan24** [comment]
@shaunagm any idea what's causing `docs-build` to fail? No clear answer from CircleCI

**Comment by shaunagm** [comment]
I don't think it's an issue with your PR. @SorenSpicknall? This and a couple other open PRs are failing with this same error `Error: [Errno 2] No such file or directory: '/home/circleci/project/venv/bin/python3'`

**Comment by SorenSpicknall** [comment]
> I don't think it's an issue with your PR. @SorenSpicknall? This and a couple other open PRs are failing with this same error `Error: [Errno 2] No such file or directory: '/home/circleci/project/venv/bin/python3'`

I'm working on this now; definitely looks like a CircleCI config breakage. If we can't figure it out in the next day or so, we may just merge some of these newer PRs without the additional testing step in order to not pause ongoing work, since the PRs generally test green aside from the CI issue.

File: parsons/scytl/scytl.py
Comment (agreenspan24) [reply parsons/scytl/scytl.py:75]: Sorry, that was a bit of a word salad. Does it make sense to you to split up the state-wide election and county-wide election into two separate connectors, since they have slightly different functions and outputs?

Separately, if we're using this for polling, to prevent fetching unnecessary results, does it make sense to you to store the previous version in the class or return it with the results and have them pass it back in the function?

**Comment by shaunagm** [comment]
@SorenSpicknall I know you submitted a PR to fix the CI issue (https://github.com/move-coop/parsons/pull/752), what do we need to do to get this PR passing? Do we need to rebase it onto the updated main branch?

**Comment by SorenSpicknall** [comment]
> @SorenSpicknall I know you submitted a PR to fix the CI issue (#752), what do we need to do to get this PR passing? Do we need to rebase it onto the updated main branch?

Yes, changes from `main` will need to be merged in or the PR will need to be recreated from scratch based on that newly updated `main` branch. Sorry about the annoyance of that, @agreenspan24 - you just got unlucky with timing of a weird cache issue with our CI.

**Comment by shaunagm** [comment]
Let me know if you need help rebasing, @agreenspan24. It can be tricky.

**Comment by shaunagm** [comment]
Hi @agreenspan24! Just checking in. I know a lot of folks are busy with the election, so I wanted to offer a few different options here:

(a) I can rebase your PR for you (I should be able to as a maintainer)
(b) we can pair and I can walk you through the process (don't know if you've rebased before, but it can be intimidating)
(c) we can just wait for you to have a chance to do it yourself

Totally up to you!

**Comment by agreenspan24** [comment]
> Hi @agreenspan24! Just checking in. I know a lot of folks are busy with the election, so I wanted to offer a few different options here:
> 
> (a) I can rebase your PR for you (I should be able to as a maintainer) (b) we can pair and I can walk you through the process (don't know if you've rebased before, but it can be intimidating) (c) we can just wait for you to have a chance to do it yourself
> 
> Totally up to you!

@shaunagm so sorry I missed this request, I can find some time in the next few days to rebase

**Comment by agreenspan24** [comment]
@shaunagm all done!

**Review by shaunagm — APPROVED** [review]

**Comment by shaunagm** [comment]
Thank you so much for the effort you put into this PR! Excited to finally get to merge it. :)

**Pull request merged**
